site/app/controllers/NotificationController.php

```diff
@@ -0,0 +1,101 @@
+<?php
+
+namespace app\controllers;
+
+use app\libraries\Core;
+use app\models\Notification;
+use app\controllers\AbstractController;
+use app\libraries\Output;
+
+/**
+ * Class NotificationController
+ *
+ */
+class NotificationController extends AbstractController {
+
+	const NOTIFICATION_SELECTIONS = [
+	    'merge_threads',
+        'all_new_threads',
+        'all_new_posts',
+        'all_modifications_forum',
+        'reply_in_post_thread',
+        'merge_threads_email',
+        'all_new_threads_email',
+        'all_new_posts_email',
+        'all_modifications_forum_email',
+        'reply_in_post_thread_email'
+    ];
+
+	public function __construct(Core $core) {
+        parent::__construct($core);
+    }
+
+    public function run() {
+        switch ($_REQUEST['page']) {
+            case 'notifications':
+                $this->notificationsHandler();
+                break;
+            case 'notification_settings':
+                $this->viewNotificationSettings();
+                break;
+        	case 'alter_notification_settings':
+        		$this->changeSettings();
+        		break;
+        }
+    }
+
+    public function notificationsHandler() {
+        $user_id = $this->core->getUser()->getId();
+        if(!empty($_GET['action'])) {
+            if($_GET['action'] == 'open_notification' && !empty($_GET['nid']) && is_numeric($_GET['nid']) && $_GET['nid'] >= 1) {
+                $metadata = $this->core->getQueries()->getNotificationInfoById($user_id, $_GET['nid'])['metadata'];
+                if(!$_GET['seen']) {
+                    $thread_id = Notification::getThreadIdIfExists($metadata);
+                    $this->core->getQueries()->markNotificationAsSeen($user_id, $_GET['nid'], $thread_id);
+                }
+                $this->core->redirect(Notification::getUrl($this->core, $metadata));
+            } else if($_GET['action'] == 'mark_as_seen' && !empty($_GET['nid']) && is_numeric($_GET['nid']) && $_GET['nid'] >= 1) {
+                $this->core->getQueries()->markNotificationAsSeen($user_id, $_GET['nid']);
+                $this->core->redirect($this->core->buildUrl(array('component' => 'notification', 'page' => 'notifications')));
+            } else if($_GET['action'] == 'mark_all_as_seen') {
+                $this->core->getQueries()->markNotificationAsSeen($user_id, -1);
+                $this->core->redirect($this->core->buildUrl(array('component' => 'notification', 'page' => 'notifications')));
+            }
```

Comment:
while here, add an `else` clause that does a trivial redirect back to the main notification page so that a user cannot potentially get a bad page here (e.g. by doing to an invalid action)